### PR TITLE
Webpack 4, react-error-overlay, react-loadable (major)

### LINKED ISSUE #122: main.js sourcemap not copied when running export
When I export my Next.js project as a static website, the main.js sourcemap isn't included. The sourcemap plugin is included in my `next.config.js` file.

### LINKED ISSUE #1428: Memory leak when refreshing during rebuild
Sometimes when I modify a file, and next.js is rebuilding, and then quickly move to the browser and refresh the page, the browser's memory quickly goes up until the computer freezes.

Browser: Firefox 52.0
Next version: next@2.0.0-beta.38

### LINKED ISSUE #2480: Bundle different modules into a single chunk
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

Users should be able to bundle different modules into one final chunk.

For example, users importing component `hello1.js` and `hello2.js` may want to import both modules at the same time:
```javascript
const Hello1 = dynamic(import('../components/hello1'))
const Hello2 = dynamic(import('../components/hello2'))
// This will request both hello1.js and hello2.js.
```

## Current Behavior

Webpack 3 allows you to do so using magic comments:
```javascript
const Hello1 = dynamic(/* webpackChunkName: "hello" */ import('../components/hello1'))
const Hello2 = dynamic(/* webpackChunkName: "hello" */ import('../components/hello2'))
// This should request only one hello.js with both modules.
```
But magic comments don't work on **nextjs**.

## Context

Sometimes a user requesting a module is going to need more modules in the future and you can bundle those in a single chunk to avoid server round trips.

## Your Environment

* Next.js version: 3.0.1-beta.5
* Environment name and version (e.g. Chrome 39, Node.js 5.4): Chrome 61, Node 6.11
* Operating System (Linux, maxOS, Windows): macOS

### LINKED ISSUE #2643: HMR doesn't work when dynamic component is loaded via regular import
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

Next should refresh dynamic imported components using HMR.

## Current Behavior

This bug depends on the place where the dynamic component is created. It works when the dynamic component is created on a page, but it doesn't work when it's created outside and imported using a regular `import x from`.

## Steps to Reproduce (for bugs)

Create a component:

```javascript
export default () => <p> I am the component</p>;
```
If you import that component and create the dynamic component on that page, HMR works:

```javascript
import dynamic from 'next/dynamic';

const DynamicComponent = dynamic(import('../components/Component.js'));

export default () => (
  <div>
    <DynamicComponent />
  </div>
);
```

But if the dynamic import is on a different file, for example `DynamicComponent.js`:

```javascript
import dynamic from 'next/dynamic';

const DynamicComponent = dynamic(import('./Component'));

export default DynamicComponent;
```

And you import it on a page with a regular import, HMR doesn't work:

```javascript
import DynamicComponent from '../components/DynamicComponent.js';

export default () => (
  <div>
    <DynamicComponent />
  </div>
);
```

The rest of the dynamic import functionalities seem to work fine even on both cases, like SSR.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    |    3.0.1-beta.18     |
| node    |      6.10   |
| OS      |     Sierra   |
| browser |    Chrome   |

### LINKED ISSUE #2894: SameLoopPromise fails when only an onRejected handler is passed to .then()
When some libraries get ahold of a Promise, they don't always add an `onFulfilled` handler, and they don't always add an `onRejected` handler using `.catch()`. Sometimes it looks like this:

`.then(undefined, onError)`

which is perfectly valid. I just experienced this error because GraphQL.js does this, for example: https://github.com/graphql/graphql-js/blob/85bfd13b5c63b4896d7643f8cd37a4f918fc5261/src/execution/execute.js#L369

But, `SameLoopPromise` always assumes that `onFulfilled` (`onResult` in its case) will be a function and tries to call it: https://github.com/zeit/next.js/blob/e13bb3f62f2c26d86cfe7bc41c0dc68456964707/lib/dynamic.js#L196

- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior
If no `onFulfilled` handler is passed, don't try calling it.

## Current Behavior
`TypeError: onResult is not a function`

## Steps to Reproduce (for bugs)
Return an `import()` result that another library attaches a handler to like `.then(undefined, onError)`. In my case: write a GraphQL resolver that returns the result of `import()`.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    | 3.2.1 |
| node    | 8.2.1 |

### LINKED ISSUE #2940: Dynamic import fails on named exports
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

Dynamically importing a named export should work like this:

```jsx
const DynamicFoo = dynamic(import('../components/Foo').then(m => m.Foo));
```

Additionally, default exports need also to be explicitly imported when following the spec – something that TypeScript needs you to do, otherwise you'll get an error.

```jsx
// Accessing m.default explicitly is mandatory when using TypeScript
const DynamicFoo = dynamic(import('../components/Foo').then(m => m.default));
```

## Current Behavior

When accessing a page with an import like above, I get the following error:

```
Path must be a string. Received undefined
TypeError: Path must be a string. Received undefined
    at assertPath (path.js:28:11)
    at join (path.js:1239:7)
    at loadChunks (/xyz/node_modules/next/dist/server/render.js:528:37)
```

(oddly, this doesn't happen when the named import is added to a hot-reloaded page)

## Context
- https://github.com/zeit/next.js/pull/2612 seems to be related to this issue
- using `m.__webpackChunkName` looks suspicious because the chunk name turns out to be `undefined` triggering above error, e.g. https://github.com/zeit/next.js/blob/master/lib/dynamic.js#L69
- next.js version: 3.2.2

### LINKED ISSUE #3165: `webpackDevMiddleware.quiet` hides useful error message when set to true by default
If there is a loader missing/not installed for example it will only say:

```
These modules were not found:

* ./pages/_document.js?entry in multi ./pages/_document.js?entry
* /Users/Vaughan/dev-live/vwp/node_modules/.registry.npmjs.org/next/3.0.6/node_modules/next/dist/pages/_error.js?entry in multi ./node_modules/.registry.npmjs.org/next/3.0.6/node_modules/next/dist/pages/_error.js?entry

To install it, you can run: npm install --save /Users/Vaughan/dev-live/vwp/node_modules/.registry.npmjs.org/next/3.0.6/node_modules/next/dist/pages/_error.js?entry
```

But when I set `quiet` to `false`, I see this:

```
ERROR in multi ./pages/_document.js?entry
Module not found: Error: Can't resolve 'babel-loader' in '/Users/Vaughan/dev-live/vwp'
 @ multi ./pages/_document.js?entry
```

...which tells me exactly how to fix it.

1. `next` should be smarter about errors, and not ask me to install absolute paths.
2. It should show all errors by default.

### LINKED ISSUE #3775: Dynamic import() fails in next 5.xx
I have been using this piece of code to asynchronously load one of heavy component on my client side. It was working like a charm:

`
const DynamicChart = dynamic(
    import('../components/LineChart/LineChart'),
    {
        loading: () => <SectionPlaceholder><Loader /></SectionPlaceholder>,
        ssr: false,
    },
);
`

Since update to v5.0.0 I get webpack error, but only when I open the parent page from frontend route. When I directly load from server, it get's loaded. When I get it once (from first load) and go back to this page on frontend, it works too.

This is the error I get:
image

I tried `import(...).then(...).catch(...)` but none of chains were triggered.

The problem is that I can't reproduce it always.

I'm also using next-routes if that matters

I assume that problem occurs because of `undefined` part in webpack. Any ideas how to debug the cause?

My package dependencies:

```
 "dependencies": {
    "babel-plugin-styled-components": "^1.5.0",
    "dotenv": "^5.0.0",
    "dotenv-webpack": "^1.5.4",
    "enzyme": "^3.3.0",
    "eslint-plugin-jest": "^21.7.0",
    "express": "^4.16.2",
    "glamor": "^2.20.40",
    "glamorous": "^4.11.4",
    "htmlescape": "^1.1.1",
    "isomorphic-fetch": "^2.2.1",
    "lodash": "^4.17.5",
    "next": "^5.0.0",
    "next-redux-wrapper": "^1.3.5",
    "next-routes": "^1.3.0",
    "polished": "^1.9.2",
    "react": "^16.2.0",
    "react-dom": "^16.2.0",
    "react-extras": "^0.5.0",
    "react-icons": "^2.2.7",
    "react-markdown": "^3.2.0",
    "react-modal": "^3.1.12",
    "react-no-ssr": "^1.1.0",
    "react-redux": "^5.0.6",
    "react-sizeme": "^2.3.6",
    "recharts": "^1.0.0-beta.10",
    "redux": "^3.7.2",
    "redux-logger": "^3.0.6",
    "redux-thunk": "^2.2.0",
    "reselect": "^3.0.1",
    "socket.io": "^2.0.4",
    "socket.io-client": "^2.0.4",
    "yargs": "^11.0.0"
  },
  "devDependencies": {
    "@storybook/addon-actions": "^3.3.12",
    "@storybook/addon-info": "^3.3.12",
    "@storybook/addon-links": "^3.3.12",
    "@storybook/addons": "^3.3.12",
    "@storybook/react": "^3.3.12",
    "@types/enzyme": "^3.1.8",
    "@types/jest": "^22.1.1",
    "@types/next": "^2.4.7",
    "@types/node": "^9.4.1",
    "@types/react": "^16.0.36",
    "@types/recharts": "^1.0.13",
    "@types/reselect": "^2.2.0",
    "@types/socket.io": "^1.4.31",
    "@types/storybook__react": "^3.0.6",
    "babel-core": "^6.26.0",
    "babel-eslint": "^7.2.3",
    "babel-plugin-recharts": "^1.1.1",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-preset-env": "^1.6.1",
    "babel-preset-es2015": "^6.24.1",
    "babel-preset-stage-0": "^6.24.1",
    "babili-webpack-plugin": "^0.1.2",
    "console-probe": "^2.0.4",
    "create-component-app": "^1.6.2",
    "eslint": "^4.17.0",
    "eslint-config-airbnb": "^16.1.0",
    "eslint-plugin-babel": "^4.1.2",
    "eslint-plugin-import": "^2.8.0",
    "eslint-plugin-jsx-a11y": "^6.0.3",
    "eslint-plugin-react": "^7.6.1",
    "eslint-plugin-styled-components-config": "0.0.2",
    "jest": "^22.1.4",
    "prop-types": "^15.6.0",
    "redux-devtools-extension": "^2.13.2",
    "webpack-bundle-analyzer": "^2.10.0"
  }
```

- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior
Component gets loaded every time

## Current Behavior
Sometimes component is not loaded and webpack tries to load something with `undefined` in name

### LINKED ISSUE #3823: Dynamic imports causes 404 in production
Currently attempting to deploy a next.js to production with dynamic imports (without using `next/dynamic` since we're importing a non-react library), and running into issues with 404s on the dynamically imported chunks.

The built folder structure looks like:
```
.next
	- app.js
	- BUILD_ID
	- 0-cd29260af35981b4adfc.js
	- ...etc
	
```
While loading the page, a javascript file at `/_next/webpack/0-cd29260af35981b4adfc.js` is loaded, pointing to the file `0-cd29260af35981b4adfc.js` to be in the `.next/chunks` directory, which it's not. Moving that file manually to the `chunks` folder fixes it. Currently, the request `404s`.

- [x ] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

The request for the dynamically loaded javascript file should not 404.

## Current Behavior

The request for the dynamically loaded javascript `404s`.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    |  5.0.0       |
| node    |   8.9.4      |
| OS      |   MacOs      |
| browser |  Chrome       |
| etc     |         |

### LINKED ISSUE #3923: Migrate to Webpack 4
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

## Context

We ❤️next.js! I made the (possibly) poor decision of migrating our (nteract's) monorepo to Webpack 4 only to notice that next.js is on webpack 3. How far off is a migration to webpack 4? My primary reason for asking is to know if I should switch us _back_ to webpack 3 or just wait on the next next version. 😄 

We have a few non-next.js apps (some loaded by a python server, one by electron) and we want to make the webpack similar using the universal webpack configurator plugins crafted for next 4. I don't want to break our next apps, so I kept those a bit separate.

### LINKED ISSUE #4100: How to use Template literals with dynamic import?
How to use Template literals with dynamic import?
`
const Component = dynamic(import(`/components/${kindOf}`))
`
In console i have error

`Module build failed: TypeError ... Cannot read property '0' of undefined`

### LINKED ISSUE #4175: Error when build page first time with dynamic import
I have a login page, after login web route to page index with `Router.push('/')`. In index page I have dynamic import some component depend on `query.page`

```
import * as React from 'react'
import * as withRedux from 'next-redux-wrapper'
import dynamic from 'next/dynamic'

import Layout from '../src/components/share/Layout'
import configStore from '../src/store/configStore.dev'

export interface Props {
  page: string
  pageProps: any
}

export interface State {}

interface DynamicBundleProps {
  page: string
  pageProps: any
}

const DynamicBundle = dynamic<DynamicBundleProps, {}>({
  modules: props => {
    const components: any = {}
    console.log(props.page)
    switch (props.page) {
      case '/lists/create': {
        components.Page = import('./lists/create')
        break
      }
      case '/lists/manage': {
        components.Page = import('./lists/manage')
        break
      }
      case '/templates/create': {
        components.Page = import('./templates/create')
        break
      }
      case '/templates/manage': {
        components.Page = import('./templates/manage')
        break
      }
      case '/campaigns/create': {
        components.Page = import('./campaigns/create')
        break
      }
      case '/campaigns/manage': {
        components.Page = import('./campaigns/manage')
        break
      }
      case '/settings': {
        components.Page = import('./settings')
        break
      }
      case '/': {
        components.Page = import('../src/components/Dashboard')
      }
      default: {
        components.Page = import('../src/components/Dashboard')
      }
    }
    console.log(components)
    return components
  },
  render: (props, { Page }) => {
    console.log(Page)
    return (
      <React.Fragment>
        <Page {...props.pageProps} />
      </React.Fragment>
    )
  },
} as any)

class IndexPage extends React.Component<Props, State> {
  constructor(props) {
    super(props)
  }

  static async getInitialProps(context) {
    const { page, ...pageProps } = context.query
    return { page, pageProps }
  }

  render() {
    let { page, pageProps } = this.props
    return (
      <Layout>
        <DynamicBundle page={page} pageProps={pageProps} />
      </Layout>
    )
  }
}

export default withRedux(configStore)(IndexPage)


```
But when I route to page index the first time after login, it build the page and then throw error before this even run, so no page is show
```
  render: (props, { Page }) => {
    console.log(Page)
    return (
      <React.Fragment>
        <Page {...props.pageProps} />
      </React.Fragment>
    )
  }
```

This is the error
nextjs_error

This error only happen on the first page build and not happen when build out production code

- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

## Expected Behavior

No error and page show normal
## Current Behavior

Throw error before render dynamic component
## Your Environment

| Tech    | Version |
|---------|---------|
| next    |  ^6.0.0-canary.3       |
| node    |  9.1.0       |
| OS      |  manjaro-linux       |
| browser |  chromium       |
| etc     |         |

### LINKED ISSUE #4194: Context provided in _app.js can't be consumed in pages in SSR
React v16.3 context provided in `_app.js` can be consumed and rendered in pages on the client, but is undefined in SSR. This causes React SSR markup mismatch errors.

Note that context can be universally provided/consumed *within* `_app.js`, the issue is specifically when providing context in `_app.js` and consuming it in a page such as `pages/index.js`.

- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

## Expected Behavior
Context provided in _app.js should be consumable in pages both on the server for SSR and when browser rendering.

## Current Behavior
Context provided in _app.js is undefined when consumed in pages for SSR. It can only be consumed for client rendering.

## Steps to Reproduce (for bugs)

In `_app.js`:

```jsx
import App, { Container } from 'next/app'
import React from 'react'
import TestContext from '../context'

export default class MyApp extends App {
  render () {
    const { Component, pageProps } = this.props
    return (
      <Container>
        <TestContext.Provider value="Test value.">
          <Component {...pageProps} />
        </TestContext.Provider>
      </Container>
    )
  }
}
```

In `context.js`:

```jsx
import React from 'react'

export default React.createContext()
```

In `pages/index.js`:

```jsx
import TestContext from '../context'

export default () => (
  <TestContext.Consumer>
    {value => value}
  </TestContext.Consumer>
)
```

Will result in:

screen shot 2018-04-23 at 2 29 29 pm

## Context

A large motivation for the `_app.js` feature is to be able to provide context persistently available across pages. It's unfortunate the current implementation does not support this basic use case.

I'm attempting to isomorphically provide the cookie in context so that [`graphql-react`](https://github.com/jaydenseric/graphql-react) `<Query />` components can get the user's access token to make GraphQL API requests. This approach used to work with separately decorated pages.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    | v6.0.0-canary.5 |
| node    | v9.11.1 |

### LINKED ISSUE #4285: HMR with TypeScript causes infinite reloading loops
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

👋 Hey everyone, I've been using next with TypeScript (was 6.0 canary, now with 6.0.1 canary) and HMR has been causing quite a lot of issues. I created an example repo to showcase it: nextjs-typescript-hrm-bug-example

Basically the default setup for HRM in nextjs means that making a change in a `*.tsx` file would cause an infinite loop with these errors

screen shot 2018-05-06 at 4 08 44 pm

Which you need to kill the server to stop.

In making the example repo, I came up with the solution in my [`next.config.js`](https://github.com/orta/nextjs-typescript-hrm-bug-example/blob/master/next.config.js):

```js
    if (dev && !isServer) {
      const HMR = config.module.rules.find(r => r.loader === "hot-self-accept-loader")
      if (HMR) {
        HMR.test = /\.(ts|tsx|js|jsx)$/
        HMR.options.extensions = /\.(ts|tsx|js|jsx)$/
      }
    }
```

Which modifies the HMR to support TS.

## Expected Behavior

This should probably be documented somewhere 👍 

Perhaps the with-typescript example could be updated to babel's TypeScript instead, and this can be added to the config file?

## Steps to Reproduce (for bugs)

```
git clone https://github.com/orta/nextjs-typescript-hrm-bug-example
cd nextjs-typescript-hrm-bug-example
yarn
yarn dev
```

Then change `index.tsx`

## Context

n/a

### LINKED ISSUE #4433: Doesn't work with multiple import
**Describe the bug**
This code causes the "handle-import" babel plugin to break:
```js
import(
  /* webpackInclude: /\.docs\.js$/, webpackChunkName: "whatever-[request]" */
  `__cwd/${path}`,
)
```

**To Reproduce**
1. Include the snippet above in a code.
2. Receive `Module build failed: TypeError: Cannot read property '0' of undefined` error from `node_modules\next\dist\server\build\babel\plugins\handle-import.js:35:30`

**Expected behavior**
I'd expect Next.js to be able to handle this kind of `import` usage, as it's [supported by webpack](https://webpack.js.org/api/module-methods/#import-). The old way to do the same, namely `require.context`, is not breaking like this (though it's breaking in other ways, as I'll bescribe in "additional context").

**System information**
 - OS: Windows 10
 - Version of Next.js: 6.0.3

**Additional context**
Related issue: https://github.com/zeit/next.js/issues/2690

This problem is directly caused by the "handle-import" plugin expecting the first argument to be a string literal (which has the `value` property). Since in my case it's a template literal, it doesn't have a `value`, only `expressions` and `quasis`.

---

For a while now I've been trying to create a wrapper around Next.js that's able to work with dynamically imported files (a tool for documentation, which includes all `*.docs.js` files in the project).

First of all, I can't use `pages/` - I want documentation files to be close to the scripts they document. I want the original structure of the source to be reflected by the structure of documentation, so replicating the directory structure in two places (original and pages) seems like an unmaintainable idea.

That's how I arrived at a solution that uses import-all.macro. The problem with that is, it's powered by babel. So when, as a user, I'm adding a new documentation file, I'll need to stop the tool, purge the cache (since all imports are inlined in a file), and restart the tool. I'm losing the file-listening capabilities of webpack-dev-server, and I need to go around the limitations of babel caching.

That's how I arrived at a solution that uses `require.context`. It _almost_ works! I can't get the SSR story correctly though - while the server renders the appropriate script, the chunk doesn't seem to be ready when `dynamic` is rendering the component. Actually, the chunk doesn't seem to be generated at all! That's why I switched to `import`, to try and see if I could get the chunk-creating machine going at least with the other way of importing. As it is now, I've spent several days on this, I'm spinning in place, and I'm determined to fix the problem.

### LINKED ISSUE #4515: Elegance of dynamic imports in TypeScript
# Bug report

## Describe the bug

This is a follow-up to https://github.com/zeit/next.js/issues/3389, which I cannot comment since the discussion is closed by the bot. The issue is to do with how dynamic imports work in TypeScript and it is still observed in Next 6.0.

## To Reproduce

The example is copy-pasted from a [comment](https://github.com/zeit/next.js/issues/3389#issuecomment-371978044) by @mohammedzamakhan:

```ts
import dynamic from "next/dynamic";

const Button = dynamic(import("../components/Button"));
```

The above code produces the following error:

```
Argument of type 'Promise<typeof "/components/Button" >' is not assignable to parameter of type 'Promise<ComponentType<{}>>'.
```

## Expected behavior

I expected `Button` to be a React component, just as when I use a non-dynamic import:

```ts
import Button from "../components/Button";
```

## Workarounds

1. Inspired by https://stackoverflow.com/a/49272129/1818285

    ```ts
    import unusedModule from "../components/Button";
    type ButtonType = typeof unusedModule;
    const Button = dynamic((import("../components/Button") as any) as Promise<ButtonType>);
    ```

2. Obtained through trial and error

    ```ts
    const Button = dynamic(import("../components/Button").then(module => module.default));
    ```

Both solutions are unintuitive, so I'm wondering if there is a more elegant way of correctly deriving component types during dynamic imports. Can it be the case that [@zeit/next-typescript](https://github.com/zeit/next-plugins/tree/master/packages/next-typescript) is suggesting a sub-optimal tsconfig (which I followed)? How to just make this notation work?

```ts
const Button = dynamic(import("../components/Button"));
``` 

## System information
 - OS: macOS
 - Version of Next.js: e.g. 6.0.3
 - @zeit/next-typescript: 1.0.1
 - tsconfig.json

```
{
  "compileOnSave": false,
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "jsx": "preserve",
    "allowJs": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "removeComments": false,
    "preserveConstEnums": true,
    "sourceMap": true,
    "skipLibCheck": true,
    "strictNullChecks": true,
    "baseUrl": ".",
    "lib": ["dom", "es2016"]
  },
  "include": ["src/**/*"]
}
```

### LINKED ISSUE #4629: Dynamic imports in _app.js do not hot reload
# Bug report
Dynamic imports in _app.js do not hot reload

## Describe the bug
When using an _app.js file, which includes a dynamic import (using `next/dynamic`), changes to the dynamically imported file will not hot reload

## To Reproduce
1. Go to [morgs32/with-dynamic-imports](https://github.com/morgs32/with-dynamic-imports/)
2. This is a variation of a pre-existing nextjs example
3. Follow steps in Readme.md

## Expected behavior
The expected behavior is changes to a dynamically imported component will hot reload (as they do in pages)

## System information
 - OS: macOS
 - Version of Next.js: 6.0.3

### LINKED ISSUE #4642: Error of dynamic component and "Loading chunk * failed"
# Bug report

## Describe the bug

each page has a dynamic component.
link to one of other pages, and the dynamic component load failed

## To Reproduce

example: https://github.com/comus/with-dynamic-import

1. rm -rf .next (clear build cache first)
2. yarn dev
3. go to localhost:3000/
4. click link "hello2"
5. error

## Expected behavior
go to every page without error

## Screenshots
https://cl.ly/1Y0g0J43273y

## System information
 - OS: mac
 - Browser (if applies) chrome
 - Version of Next.js: 6.0.3

## Additional context

none

### LINKED ISSUE #4708: ErrorBoundary useless with React 16.4 handling of getDerivedStateFromProps
# Bug report

## Describe the bug
With the new handling of `getDerivedStateFromProps`, the ErrorBoundary's state is constantly reset (error to `null`) before *every* render. Therefore it is never able to render the `ErrorReporter` and instead attempts to rerender the faulty Page/react components.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. https://github.com/morgs32/with-next-routes
2. Click on either of the two buttons to go to the `about` page

## Expected behavior
Expect the ErrorBoundary to render `next/_error.js` with the associated error and stacktrace.

## System information
 - ~~OS: [e.g. macOS, Windows]~~
 -  ~~Browser~~
 - React 16.4.1
 - Version of Next.js: 6.1.1

### LINKED ISSUE #4711: Offline story of the up-to-date detection logic
# Bug report

## Describe the bug

I believe we miss the offline story of the up-to-date detection logic. I'm working in the train, with a Li-Fi connection  (crappy).

If you are lucky, it fails fast with the following error:
capture d ecran 2018-07-01 a 19 58 35

However, if the connection lies to you (Li-Fi, it think it has internet but it doesn't) then you are stuck. The dev server refuses to start. He is waiting for a network request to complete.
After 1-2 minutes, you will get the following error:
capture d ecran 2018-07-01 a 19 51 53

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. Start the dev server without internet or with a fake internet connection
2. See error

## Expected behavior

I don't think that the up-to-date feature is important enough to harm offline development (we already have `yarn oudated` to solve this problem).

## System information
 - Version of Next.js: 6.0.4-canary.8

### LINKED ISSUE #4737: `unreachable code after return statement` warning
# Bug report

## Describe the bug

there are `unreachable code after return statement` warning in firefox (in production).

i found that it causes by utils `deprecated` function

`deprecated` compiled as
```js
  t.deprecated = function (e, t) {
    return e;   // <----------------- first return (in production)
    var n = !1,
    r = function () {
      n || (n = !0, console.error(t));
      for (var r = arguments.length, o = new Array(r), a = 0; a < r; a++) o[a] = arguments[a];
      return e.apply(this, o)
    };
    return (0, i.default) (r, e), r
  },
```

the source code is
https://github.com/zeit/next.js/blob/canary/lib/utils.js#L17

## To Reproduce

1. Go to  http://zeit.co (firefox)
2. open devtools
3. see console
4. warning here

or

1. yarn create next-app --example hello-world hello-world-app
2. cd hello-world-app
3. yarn build
4. yarn start
5. go to http://localhost:3000 (firefox)

## Expected behavior
:)

## Screenshots
https://www.dropbox.com/s/m700m4c8i4k2jru/Screenshot%202018-07-06%2004.57.10.png?dl=0
https://www.dropbox.com/s/iw43kabb2rthpjl/Screenshot%202018-07-06%2005.05.09.png?dl=0

## System information
 - OS: macOS
 - Browser (if applies) firefox 61.0
 - Version of Next.js: 6.1.1

## Additional context

**it is not very important, so i dont submit a PR, just wanna notes that there are some unnecessary warning**

### LINKED ISSUE #4783: How disable update checker?
I use next.js with Electron and every time when launch app in dev more I should accept or reject dialog from my firewall. I prefer disable this update checking at all.

According to [this part of code](https://github.com/zeit/next.js/blob/master/server/index.js#L112) it seems we haven't any abilities to disable this?

### LINKED ISSUE #4784: Not fatal but weird bug
# Bug report

When running develop instance. I see it building and showing the pages correctly. Then after a while it shows. "> Disposing inactive page(s).... pagename" at that point things still work pretty well but the browser front end shows weird errors with partially invalid javascript. Words truncate, dangling "return {" at the end of files and so on. Something is partially messing up the cached code. Maybe there is a timing issue where it sends a file to the browser while it's compiling a different page into the same code. Seems weird.

## To Reproduce

All I have to do is run "npm run develop" go to a page with graphql and let it sit for a while. All of a sudden I see weird errors in the browser development tools.

## Expected behavior

This shouldn't really happen.

## Screenshots

image
image

In this case there is part of a list and part of the object overwritten.

## System information
 - OS: Mac OS X 10.13.5 (17F77)
 - Browser (if applies) Chrome  67.0.3396.99 (Official Build) (64-bit)
 - Version of Next.js: next.js v6.1.1

## Additional context

Like I said it's not a show stopped, but it should probably be fixed. It scared me a bit for a while :)

### PR DESCRIPTION
**Pull request by timneutkens**
Fixes #4783
Fixes #2643
Fixes #3775
Fixes #4515
Fixes #3923
Fixes #4433 
Fixes #4100 
Fixes #4642
Fixes #3823
Fixes #4708
Fixes #4285
Fixes #4194
Fixes #4784 
Fixes #4737 
Fixes #4711
Fixes #3165
Fixes #2480
Fixes #1428 
Fixes #2894
Fixes #4629
Fixes #2940
Fixes #4175
Fixes https://github.com/zeit/next-plugins/issues/122
Implements a few fixes for #4613

Figured out the chunk runtime merging issue described in #4108

New pull request as the codebase since the initial pull request has shifted quite a bit (folders moved etc).

What's left to do:

- [x] Failing tests (not all of these are failing, keeping a list of tests I checked and fixed already (or weren't broken))
  - [x] basic
  - [x] static
  - [x] app-aspath
  - [x] app-document
  - [x] babel
  - [x] config
  - [x] dist-dir
  - [x] custom-server
  - [x] ondemand
  - [x] production
  - [x] with-router
- [x] Correct split chunk configuration
  - [x] commons.js for modules used in 50% of all pages
  - [x] hashing of main.js and runtime.js
- [x] Upgrade uses of old webpack plugin hooks system
- [x] Replace custom error overlay with `react-error-overlay`, which improves DX + gives more accurate error locations.
  - [x] Make react-error-overlay work with errors coming from SSR
  - [x] Reload the page whenever an error is fixed
- [x] `next/dynamic` replaced by `react-loadable` under the hood. No changes in how it currently works. Only 1 specific edge case has a breaking change, see `breaking` below

### Major

- Node.js 8+ is now required.
- Webpack 4, all plugins rewritten to use the Webpack 4 API
- Re-implemented `next/dynamic` using `react-loadable` under the hood, this means that you still use `next/dynamic` but underneath it will use react-loadable instead of the previous custom solution.
- Implemented react-error-overlay / new hot-dev-client implementation (overhauled error rendering in development)
- Share module instances initiated by webpack between pages on the server side, allows for React 16.3 context between _app.js and pages and basically any other requirement for sharing module instances on the server side. I can also imagine this greatly reduces memory usage / improves speed in a serverless environment too.

### Minor

- Removed 2 "webpack specific" routes. These files are now put into `.next/static`: `.next/static/webpack` (hot update files) and `.next/static/chunks` (dynamic import files)
- Added [Webpackbar](https://github.com/nuxt/webpackbar)

### Patches

- Make sure `false` is not rendered on non _error.js pages
- Updated a lot of tests to reset their state on failure (close browser, reset file) 
- `_document.js` is only added to server compilation, while keeping HMR for it
- Figured out a bugfix for https://github.com/webpack/webpack/issues/6598
- Fixed dead code elimination for the `deprecated` method: [#4737](https://github.com/zeit/next.js/issues/4737)

## Breaking

- Node.js 8+ is now required.
- `next/dynamic` when providing `modules` as per https://github.com/zeit/next.js#4-with-multiple-modules-at-once `props` are no longer passed in as the function gets pre-loaded on the server and can't depend on props.

## After the canary release

- [ ] config tests failing on @zeit/next-css because of webpack 4, temporarily disabled.
- [ ] Update next-plugins
- [ ] Add modules imported in _app.js to commons as _app.js is loaded in every page. [Question on `webpack/webpack` for advice](https://github.com/webpack/webpack/issues/7556#issuecomment-406813394)

### CODE REVIEW

**Comment by stramel** [comment]
@timneutkens What are your Node.js supported versions? LTS is now 8. Curious if we need the 6 & 7 checks? 

BTW great work! Can't wait to try out webpack 4!

**Comment by timneutkens** [comment]
All tests run on Node.js 6, but it's recommended to use the latest LTS/later version of Node.js

**Comment by stramel** [comment]
@timneutkens not sure if you saw this about the minimizer https://twitter.com/greweb/status/1013122437884542978

**Comment by stramel** [comment]
@timneutkens I believe this will also fix #4100 ?

**Comment by timneutkens** [comment]
@stramel 😌 thanks!

Will check the minimizer out once all tests pass 👍

File: build/webpack/plugins/react-loadable-plugin.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+function buildManifest (compiler, compilation) {
+  let context = compiler.options.context
+  let manifest = {}
+
+  compilation.chunks.forEach(chunk => {
+    chunk.files.forEach(file => {
+      chunk.forEachModule(module => {
+        let id = module.id
+        let name = typeof module.libIdent === 'function' ? module.libIdent({ context }) : null
+        // If it doesn't end in `.js` Next.js can't handle it right now.
```
Comment (jaydenseric) [inline build/webpack/plugins/react-loadable-plugin.js]: Why not `.mjs`? Webpack v4 in fact prefers it.

Comment (timneutkens) [reply build/webpack/plugins/react-loadable-plugin.js]: This is the output filename, not the input filename.

**Comment by mrtkrcm** [comment]
can't wait to get this done! great work Tim 🙏 as always

**Review by Skaronator — COMMENTED** [review]
also check that `/index.js` is the same as `/index`.

**Comment by timneutkens** [comment]
@Skaronator `page` => page path, meaning it will never be `.js`.

**Comment by Skaronator** [comment]
Alright, that question just came in my mind as I saw your latest commit.

**Comment by AlexOrtizzz** [comment]
I hope and wait.

**Review by arunoda — APPROVED** [review]
LGTM

**Comment by janjarfalk** [comment]
Oh lordy!
Great work Tim!

**Comment by albinekb** [comment]
Freaking awesome work! 😍

**Comment by morgs32** [comment]
Congratulations. Small step for man...

**Comment by tomevans18** [comment]
Great work on this.
Lots of fixes.
Are you looking to release this as "6.1.1-canary.3"? Just a little desperate for the dynamic imports.

**Comment by pantharshit00** [comment]
Ya need this ASAP cause SSR is breaking on my app if I am trying to use context

**Comment by mariosanchez** [comment]
Good job! Very happy to have Webpack 4 arround the corner 👍

**Comment by jesstelford** [comment]
It looks like **Webpack 4 is included in [`6.1.1-canary.3`](https://github.com/zeit/next.js/releases/tag/6.1.1-canary.3)** :tada:

**Pull request merged**